internal/proxy/reverse_proxy.go

```diff
@@ -0,0 +1,161 @@
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	OpenshiftGatewayTimeout = 5 * time.Second
+)
+
+type reverseProxy struct {
+	redirectURL     url.URL
+	logger          *log.Entry
+	responseTimeout time.Duration
+}
+
+type directorFunc func(req *http.Request)
+
+func director(target *url.URL) directorFunc {
+	// targetQuery := target.RawQuery
+	return func(req *http.Request) {
+		log.Warnf(" > RevProxy: CALLING Director  | target:  %v | req: %q ", target, req.URL)
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+		log.Infof(" > RevProxy: Director %v", req.URL)
+	}
+
+}
+
+func newLoggedReverseProxy(target *url.URL) *httputil.ReverseProxy {
+	// copied from:
+
+	transport := &http.Transport{
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			// log.Println(" > RevProxy: CALLING Proxy")
+			return http.ProxyFromEnvironment(req)
+		},
+
+		Dial: func(network, addr string) (net.Conn, error) {
+			// log.Println(" > RevProxy: CALLING Dial")
+
+			dialer := &net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}
+
+			conn, err := dialer.Dial(network, addr)
+			if err != nil {
+				log.Infof(" > RevProxy: Dial - Error: %v", err)
+			}
+			return conn, err
+		},
+
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+
+	return &httputil.ReverseProxy{
+		Director:  director(target),
+		Transport: transport,
+		// ModifyResponse: modifyResponse,
+	}
+}
+
+type responseRecorder struct {
+	http.ResponseWriter
+	statusCode  int
+	err         error
+	wroteHeader bool
+	log         *log.Entry
+}
+
+func newResponseRecorder(w http.ResponseWriter, logger *log.Entry) *responseRecorder {
+	return &responseRecorder{
+		ResponseWriter: w,
+		log:            logger.WithField("response_recorder", "true"),
+		statusCode:     http.StatusOK,
+		err:            nil,
+		wroteHeader:    false,
+	}
+}
+
+func (rr *responseRecorder) WriteHeader(code int) {
+	rr.log.Infof("Write Header called with %d", code)
+
+	if rr.wroteHeader {
+		rr.log.Warnf("Multiple Write Header %d called; SKIPPED", code)
+		return
+	}
+
+	rr.statusCode = code
+	rr.wroteHeader = true
```

Comment:
no, because we don't want to send the actual status code to the user if it times out i.e. `http.StatusBadGateway` if it times out, then we capture the actual status in rr.statusCode, set the `err` and `return`